rules/prefer-error-cause.js

```diff
@@ -0,0 +1,319 @@
+	const errorArgumentIdentifier = parameter?.name;
+	const throwStatementArgument = throwStatement.argument;
+
+	// Filter rethrowing the error itself.
+	// `try {} catch (err) { throw err; }`
+	// `promise.then(undefined, (err) => { throw err; })`
+	// `promise.catch(err => { throw err; })`
+	if (errorArgumentIdentifier && errorArgumentIdentifier === throwStatementArgument.name) {
+		return;
+	}
+
+	// Report none identifier parameter
+	// `try {} catch ({error}) {}`
+	// `promise.then(undefined, ({error}) => {})`
+	// `promise.catch({error} => {})`
+	if (parameter && parameter.type !== 'Identifier') {
+		context.report({
+			node: catchBlock,
+			messageId: ERROR,
+		});
+		return;
+	}
+
+	const nodesToFixCandidates = getAllNodesToFix({catchBlock, context, throwStatement});
+
+	const nodesToFix = nodesToFixCandidates.filter(({node}) =>
+		!hasCauseProperty(node),
+	);
+
+	for (const nodeToFix of nodesToFix) {
+		context.report({
+			node: catchBlock,
+			messageId: ERROR,
+			fix(fixer) {
+				return fix({
+					fixer,
+					context,
+					catchBlock,
+					statementToFix: nodeToFix.statementToFix,
+					errorConstructorLastArgument: nodeToFix.node,
+					errorArgumentIdentifier,
+				});
+			},
+		});
+	}
+};
+
+const isThenMethod = callExpression =>
+	callExpression?.optional !== true
+	&& callExpression?.arguments?.length === 2
+	&& callExpression?.arguments[0].type !== 'SpreadElement'
+	&& callExpression?.arguments[1].type !== 'SpreadElement'
+	&& callExpression?.callee.type === 'MemberExpression'
+	&& callExpression?.callee.optional !== true
+	&& callExpression?.callee.computed !== true
+	&& callExpression?.callee.property.type === 'Identifier'
+	&& callExpression?.callee.property.name === 'then';
+
+const isCatchMethod = callExpression =>
+	callExpression?.optional !== true
+	&& callExpression?.arguments?.length === 1
+	&& callExpression?.arguments[0].type !== 'SpreadElement'
+	&& callExpression?.callee.type === 'MemberExpression'
+	&& callExpression?.callee.optional !== true
+	&& callExpression?.callee.computed !== true
+	&& callExpression?.callee.property.type === 'Identifier'
+	&& callExpression?.callee.property.name === 'catch';
+
+/** @param {import('eslint').Rule.RuleContext} context */
+const create = context => ({
+	[throwStatementSelector](node) {
+		const catchBlock
+			= getMatchingAncestorOfType(node, 'CatchClause')
+			|| getMatchingAncestorOfType(node, 'CallExpression', parent => isThenMethod(parent) || isCatchMethod(parent));
+
+		if (!catchBlock) {
+			return;
+		}
+
+		if (catchBlock.type === 'CatchClause') {
+			handleCatchBlock({
+				catchBlock,
+				context,
+				parameter: catchBlock.param,
+				throwStatement: node,
+			});
+		} else {
+			const functionExpression = catchBlock.callee.property.name === 'then' ? catchBlock.arguments[1] : catchBlock.arguments[0];
+			const functionInnerBlockStatements = functionExpression.body?.body;
+			if (!functionInnerBlockStatements) {
+				return;
+			}
```

Comment:
Needs a test for this for code coverage.

Reply:
> Needs a test for this for code coverage.

I wonder if these statements are moot.
I think maybe `FunctionExpression` without 'body(`BlockStatement`).body' is not possible since it causes parsing error.